lib/spack/spack/externals.py

```diff
@@ -0,0 +1,349 @@
+import re
+import uuid
+import warnings
+from typing import Any, Callable, Dict, List, NamedTuple, Tuple, Union
+
+from spack.vendor.typing_extensions import TypedDict
+
+import spack.archspec
+import spack.deptypes
+import spack.repo
+import spack.spec
+from spack.error import SpackError
+
+
+class DependencyDict(TypedDict, total=False):
+    id: str
+    spec: str
+    deptypes: spack.deptypes.DepTypes
+    virtuals: str
+
+
+class ExternalDict(TypedDict, total=False):
+    """Dictionary representation of an external spec."""
+
+    spec: str
+    prefix: str
+    modules: List[str]
+    extra_attributes: Dict[str, Any]
+    id: str
+    dependencies: List[DependencyDict]
+    required_target: str
+
+
+def node_from_dict(external_dict: ExternalDict) -> spack.spec.Spec:
+    """Returns an external spec node from a dictionary representation."""
+    extra_attributes = external_dict.get("extra_attributes", {})
+    result = spack.spec.Spec(
+        # Allow `@x.y.z` instead of `@=x.y.z`
+        str(spack.spec.parse_with_version_concrete(external_dict["spec"])),
+        external_path=external_dict.get("prefix"),
+        external_modules=external_dict.get("modules"),
+    )
+    if not result.versions.concrete:
+        raise ExternalSpecError(
+            f"The external spec '{external_dict['spec']}' doesn't have a concrete version."
+        )
+
+    result.extra_attributes = extra_attributes
+    if "required_target" in external_dict:
+        result.constrain(f"target={external_dict['required_target']}")
+    return result
+
+
+def complete_architecture(node: spack.spec.Spec) -> None:
+    """Completes a node with architecture information."""
+    if node.architecture:
+        if not node.architecture.target:
+            node.architecture.target = spack.archspec.HOST_TARGET_FAMILY
+        node.architecture.complete_with_defaults()
+    else:
+        node.constrain(spack.spec.Spec.default_arch())
+        node.architecture.target = spack.archspec.HOST_TARGET_FAMILY
+
+    for flag_type in spack.spec.FlagMap.valid_compiler_flags():
+        node.compiler_flags.setdefault(flag_type, [])
+
+
+def complete_variants_and_architecture(node: spack.spec.Spec) -> None:
+    """Completes a node with variants and architecture information."""
+    complete_architecture(node)
+    pkg_class = spack.repo.PATH.get_pkg_class(node.name)
+    variants_dict = pkg_class.variants.copy()
+
+    progress = True
+    while progress:
+        progress = False
+        current_keys = list(variants_dict.keys())
+        for key in current_keys:
+            if not node.satisfies(key):
+                continue
+            applicable_variants = variants_dict.pop(key)
+            for v in applicable_variants.values():
+                if not node.satisfies(f"{v.name}=*"):
+                    # Cannot use Spec.constrain, because we lose information on the variant type
+                    node.variants[v.name] = v.make_default()
+            progress = True
```

Comment:
Minor tweaks to make this clearer:
```suggestion
    variants_dict = pkg_class.variants.copy()
    changed = True
    
    while variants_dict and changed:
        changed = False
        items = list(variants_dict.items())  # copy b/c loop modifies dict
        
        for when, variants_by_name in items:
            if not node.satisfies(when):
                continue
            variants_dict.pop(when)
            for name, vdef in variants_by_name.items():
                if name not in node.variants:
                    # Cannot use Spec.constrain, because we lose information on the variant type
                    node.variants[name] = vdef.make_default()
            changed = True
```

I don't think this is strictly necessary for correctness:

```python
    while variants_dict and changed:
```

but it makes it clearer what the loop is doing, and it avoids a needless last iteration (and copy) if you exhaust variants_dict *and* change something at the same time.